addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -615,3 +353,298 @@ def action_see_accrual_plans(self):
+            ('state', 'in', ['confirm', 'validate1', 'validate']),
+        ]
+        if self.env.context.get("ignore_future"):
+            leaves_domain.append(('date_from', '<=', date))
+        leaves = self.env['hr.leave'].search(leaves_domain)
+        leaves_per_employee_type = defaultdict(lambda: defaultdict(lambda: self.env['hr.leave']))
+        for leave in leaves:
+            leaves_per_employee_type[leave.employee_id][leave.holiday_status_id] |= leave
+
+        allocations = self.env['hr.leave.allocation'].with_context(active_test=False).search([
+            ('employee_id', 'in', employees.ids),
+            ('holiday_status_id', 'in', self.ids),
+            ('state', '=', 'validate'),
+        ]).filtered(lambda al: al.active or not al.employee_id.active)
+        allocations_per_employee_type = defaultdict(lambda: defaultdict(lambda: self.env['hr.leave.allocation']))
+        for allocation in allocations:
+            allocations_per_employee_type[allocation.employee_id][allocation.holiday_status_id] |= allocation
+
+        # allocation_leaves_consumed is a tuple of two dictionnaries.
+        # 1) The first is a dictionary to map the number of days/hours of leaves taken per allocation
+        # The structure is the following:
+        # - KEYS:
+        # allocation_leaves_consumed
+        #  |--employee_id
+        #      |--holiday_status_id
+        #          |--allocation
+        #              |--virtual_leaves_taken
+        #              |--leaves_taken
+        #              |--virtual_remaining_leaves
+        #              |--remaining_leaves
+        #              |--max_leaves
+        #              |--accrual_bonus
+        # - VALUES:
+        # Integer representing the number of (virtual) remaining leaves, (virtual) leaves taken or max leaves for each allocation.
+        # leaves_taken and remaining_leaves only take into account validated leaves, while the "virtual" equivalent are
+        # also based on leaves in "confirm" or "validate1" state.
+        # The unit is in hour or days depending on the leave type request unit
+        # 2) The second is a dictionary mapping the remaining days per employee and per leave type that are either
+        # not taken into account by the allocations, mainly because accruals don't take future leaves into account.
+        # This is used to warn the user if the leaves they takes bring them above their available limit.
+        # - KEYS:
+        # allocation_leaves_consumed
+        #  |--employee_id
+        #      |--holiday_status_id
+        #          |--to_recheck_leaves
+        #          |--date_remaining_days
+        #          |--exceeding_duration
+        # - VALUES:
+        # "to_recheck_leaves" stores every leave that is not yet taken into account by the "allocation_leaves_consumed" dictionary.
+        # "date_remaining_days" represents the excess amount that somehow isn't taken into account by the first dictionary.
```

Comment:
How can it not be taken into account by the first dictionary?

Reply:
Because when we associate days from a leave to allocations, we carry to the next valid allocation the days of the leave that have been left to the next allocation. If that amount is carried over to the next allocation while we're at the last allocation in the list, we are left with remaining days that have not been linked to any allocation. That value will be used in the future to comply with the negative amounts depending on how it's implemented.